spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,146 @@ cdef class Doc:
                 retokenizer.merge(span, attributes[i])
 
+    def from_json(self, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        ### Token-level properties ###
+
+        words = []
+        # Map annotation type IDs to their string equivalents.
+        annot_types = {t: self.vocab.strings[t].lower() for t in (POS, HEAD, DEP, LEMMA, TAG, MORPH)}
+        token_annotations_reqs = {
+            annot_type_id: annot_type_id if annot_type_id != HEAD else DEP for annot_type_id in annot_types
+        }
+        token_annotations = {annot_type_id: None for annot_type_id in annot_types}
+
+        # Gather token-level properties.
+        for token in doc_json["tokens"]:
+            words.append(doc_json["text"][token["start"]:token["end"]])
+            for annot_type_id in token_annotations:
+                # Assuming all tokens have exactly the same set of attributes.
+                if annot_types[token_annotations_reqs[annot_type_id]] in token:
+                    if token_annotations[annot_type_id] is None:
+                        token_annotations[annot_type_id] = []
+                    token_annotations[annot_type_id].append(token[annot_types[annot_type_id]])
+
+        # Assert all tokens have exactly the same set of attributes.
+        inconsistent_props: Set[str] = set()
+        for annot_type_id in token_annotations:
+            if token_annotations[annot_type_id]:
+                if len(token_annotations[annot_type_id]) != len(words):
+                    inconsistent_props.add(annot_types[annot_type_id])
+        if len(inconsistent_props):
+            raise ValueError(Errors.E1039.format(inconsistent_props=inconsistent_props))
+
+        # Initialize doc instance.
+        start = 0
+        cdef const LexemeC* lex
+        cdef str orth_
+        cdef bint has_space
+        reconstructed_words, spaces = get_words_and_spaces(words, doc_json["text"])
+        assert words == reconstructed_words
+
+        for i, token in enumerate(doc_json["tokens"]):
+            has_space = spaces[i]
+            orth_ = doc_json["text"][token["start"]:token["end"]]
+            lex = self.vocab.get(self.mem, orth_)
+            self.push_back(lex, has_space)
+
+        # Set remaining token-level attributes via Doc.from_array().
+        if token_annotations[HEAD] is not None:
+            token_annotations[HEAD] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations[HEAD])
+            ]
+        if token_annotations[DEP] is not None:
+            deps = []
+            for dep in token_annotations[DEP]:
+                if dep is None:
+                    raise ValueError(Errors)
+                deps.append(dep)
+            token_annotations[DEP] = deps if len(deps) else None
```

Comment:
I'm not sure we'll need this exact check in the end, but I think you can simplify this to just check for `None`:

```suggestion
        if DEP in token_annotations[DEP]:
            if None in token_annotations[DEP]:
                raise ValueError(Errors.E###)
```